adding stale-nat-eip.sh for review

Scotchman0 opened the pull request.

Scotchman0 committed:
adding stale-nat-eip.sh for review

Scotchman0 commented on OCPBUGS-50709/stale-nat-eip.sh:
```diff
@@ -0,0 +1,38 @@
+            echo "suuid: $suuid"
+            echo "slr: $slr"
+            #the below command string will remove any stale NAT entries that do not match existing/expected egressIP entries. comment it out in order to log only
+            oc -n openshift-ovn-kubernetes exec -it $localpod -c northd -- ovn-nbctl remove logical_router ${slr} nat ${suuid} ; echo removed nat ${suuid} from logical router ${slr}
```
This line needs some work - currently gives an exit 1 on execution for the `ovn-nmctl remove` command but does echo the suuid and slr values properly. We may need to migrate it out of the nested loop and into it's own functional call later.

Scotchman0 replied:
figured out why this is failing - need to variable expand before the container is called - modifying it...

tssurya commented on OCPBUGS-50709/stale-nat-eip.sh:
```diff
@@ -0,0 +1,38 @@
+     localpod=$(oc get pod -n openshift-ovn-kubernetes -o wide | grep -v NAME | grep -w "$nodeName" | grep ovnkube-node | awk {'print $1'})
+     echo "localpod:"
+     echo $localpod
+     listOfNats=$(oc -n openshift-ovn-kubernetes exec -it $localpod -c northd -- ovn-nbctl --format=csv --column "_uuid, external_ids, external_ip" find nat external-ids:\"name\"!=\"\")
```
I'd say change the container to be NBDB? the database here and everywhere else. Northd might have access to NBDB but better to change all actions where you execute nbctl to be on the nbdb itself

Scotchman0 replied:
changed

tssurya commented on OCPBUGS-50709/stale-nat-eip.sh:
```diff
@@ -0,0 +1,38 @@
+for nodeName in $(oc get nodes | grep -v NAME | awk {'print $1'}); do
+     echo "=========="
+     echo $nodeName
+     ips=$(oc get egressip | awk {'print $2'} | grep -v "EGRESSIPS")
```
can we also print the ips here?

tssurya replied:
ips = $(kubectl get egressip -o json | jq -r '.items[] | .status.items[]? | select(.node == "ovn-worker") | .egressIP')

This is what we wanted to do right?

Scotchman0 replied:
added IP print , the oc query is similar to the jq output and doesn't require any additional plugins depending on customer env (which might not have jq installed).

tssurya commented on OCPBUGS-50709/stale-nat-eip.sh:
```diff
@@ -0,0 +1,38 @@
+     for nat in ${listOfNats[@]}; do
+        echo "nat being evaluated:" 
+        echo "$nat"
+        if grep -q "$ips" <<< *"$nat"*
```
can we also print the values being compared just for debugging purposes?
I am not a bash expert but what exactly does this match do?
Is it a "is contains match" ? 
also I don't see your get EIPs filtering based on this node? -> so how will we remove the stale NATs?

Scotchman0 replied:
`-q` is currently silencing the report of the check, but we are passing the string `$nat` as an input to grep and checking if the string `$ips` is found within the input string. The command will return true or false as appropriate and that dictates our flow. If we find a match, we ignore it (because it has an expected egressIP address). If it doesn't have an expected egressIP we then action on it as a stale entry. 

`<<<` is a "here-string" function. Effectively we're passing the content of the string on the right to the command being executed on the left, which in this case is executing a grep for the content encapsulated in `$ips` against the presented string from `$nat`

This `$nat` list is pre-sorted to only show us NATs that include non-null content in the `external_id` field where `name` is defined, and the content value following `name=` is not `null`. (`external-ids: name != ""`) --> escaped below:

`listOfNats=$(oc -n openshift-ovn-kubernetes exec -it $localpod -c northd -- ovn-nbctl --format=csv --column "_uuid, external_ids, external_ip" find nat external-ids:\"name\"!=\"\")`

I'll see if we can add a value comparison output for debug.

tssurya commented on OCPBUGS-50709/stale-nat-eip.sh:
```diff
@@ -0,0 +1,38 @@
+        else
+            echo "match not found, stale nat"
+            suuid=$(echo $nat | awk -F ',' {'print $1'})
+            slr=$(oc -n openshift-ovn-kubernetes exec -it $localpod -c northd -- ovn-nbctl --bare --column=_uuid,nat find logical_router | grep -B1 $suuid | awk {'print $1'} | head -n 1)
```
we only want to always target the GR here...so its usually GR_nodename syntax

tssurya commented on OCPBUGS-50709/stale-nat-eip.sh:
```diff
@@ -0,0 +1,38 @@
+            echo "match not found, stale nat"
+            suuid=$(echo $nat | awk -F ',' {'print $1'})
+            slr=$(oc -n openshift-ovn-kubernetes exec -it $localpod -c northd -- ovn-nbctl --bare --column=_uuid,nat find logical_router | grep -B1 $suuid | awk {'print $1'} | head -n 1)
+            echo "suuid: $suuid"
```
suuid is NAT's uuid?
nit: move it to above L31.

tssurya commented on OCPBUGS-50709/stale-nat-eip.sh:
```diff
@@ -0,0 +1,38 @@
+            suuid=$(echo $nat | awk -F ',' {'print $1'})
+            slr=$(oc -n openshift-ovn-kubernetes exec -it $localpod -c northd -- ovn-nbctl --bare --column=_uuid,nat find logical_router | grep -B1 $suuid | awk {'print $1'} | head -n 1)
+            echo "suuid: $suuid"
+            echo "slr: $slr"
```
is this the UUID of the router? I think we can use the name

Scotchman0 replied:
I have modified the variable names from `$suuid` and `$slr` to `nat_uuid` and `router_uuid` respectively for clarity

tssurya commented on OCPBUGS-50709/stale-nat-eip.sh:
```diff
@@ -0,0 +1,38 @@
+            echo "suuid: $suuid"
+            echo "slr: $slr"
+            #the below command string will remove any stale NAT entries that do not match existing/expected egressIP entries. comment it out in order to log only
+            oc -n openshift-ovn-kubernetes exec -it $localpod -c northd -- ovn-nbctl remove logical_router ${slr} nat ${suuid} ; echo removed nat ${suuid} from logical router ${slr}
```
let's have a dry-run which prints the whole NAT about to be deleted first i.e all the fields not just the UUID so that we can then have someone manually compare that to the customer's case and then execute the actual deletion.

Scotchman0 replied:
done - default behavior now dry-runs, must call with --clean to force nat removal

tssurya commented on OCPBUGS-50709/stale-nat-eip.sh:
```diff
@@ -0,0 +1,38 @@
+            echo "suuid: $suuid"
+            echo "slr: $slr"
+            #the below command string will remove any stale NAT entries that do not match existing/expected egressIP entries. comment it out in order to log only
+            oc -n openshift-ovn-kubernetes exec -it $localpod -c northd -- ovn-nbctl remove logical_router ${slr} nat ${suuid} ; echo removed nat ${suuid} from logical router ${slr}
```
ovn-nbctl: 'remove' command requires at least 4 arguments

@dceara remind me deleting the NAT from the router will auto remove dangling unreferenced NATs correct?

dceara replied:
@tssurya Yes, the NB.NAT table is defined as "isRoot: false" so unreferenced NATs will be garbage collected.

Scotchman0 replied:
error message re: 4 args is present because it's being executed errantly on a null nat entry, this exception has been resolved.

Scotchman0 commented:
@CourtCourt521 looping in for visibility + collab

Scotchman0 committed:
encased the script in a function block to be called with CASE handling, added improved comments explaining logical flow, added option to remove nats only with --clean. todo: add --debug handling + dump nat table options

Scotchman0 committed:
added dump-tables which should provide output for cross-reference and tarball log function, --clean option still not sufficiently tested do not use

Scotchman0 commented:
Currently the script needs implementation on the `--debug` arg, and testing on the `--clean` arg, but I've tested it repeatedly otherwise and it does dump the logs and NAT tables successfully when run without arguments or with, so we can easily eyeball the configuration. If we want to make life simple, I can just remove the --clean argument entirely, we can put a KCS together for this to run it explicitly as a data-gathering tool only, and we can redirect to an OVNKube DB rebuild process for impacted hosts as "clean" step process. I presume also that we'll want customers to upgrade anyway rather than periodically flushing the NAT entries, and I can list in private notes a one-off clean command that can be run in emergencies. @tssurya let me know your thoughts, might be better to just have this be data-gather only

Scotchman0 committed:
removed --clean flag option and --debug, instead added verbosity directly to the output and converted this to a tool for gathering current state only - KCS to be created for how to purge and will be tied to OCPBUGS-50709

Scotchman0 commented:
I've removed case handling, removed --clean argument, removed --debug argument (just prints everything now) and have changed the scope of this script to be explicitly a data-gathering tool with no implementation expectations about removing stale nats after finding them. KCS has been created at https://access.redhat.com/solutions/7110252 (currently private) to detail the steps. In limiting the scope to data-gathering only I think we should be safe enough. re-tagging for review

Scotchman0 committed:
updated to include a fixed ip gather command using jq to ensure we catch multi-egressIP entries, and added an early-exit if statement to confirm jq is installed to parse properly on execute

Scotchman0 committed:
minor correction under dump-tables function to correct ip variable to jq output as well
